File: app/src/main/java/org/thoughtcrime/securesms/database/MegaphoneDatabase.java
package org.thoughtcrime.securesms.database;

import android.app.Application;
import android.content.ContentValues;
import android.database.Cursor;

import androidx.annotation.NonNull;

import net.sqlcipher.database.SQLiteDatabase;
import net.sqlcipher.database.SQLiteOpenHelper;

import org.signal.core.util.logging.Log;
import org.thoughtcrime.securesms.crypto.DatabaseSecret;
import org.thoughtcrime.securesms.crypto.DatabaseSecretProvider;
import org.thoughtcrime.securesms.database.model.MegaphoneRecord;
import org.thoughtcrime.securesms.megaphone.Megaphones.Event;
import org.thoughtcrime.securesms.util.CursorUtil;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * IMPORTANT: Writes should only be made through {@link org.thoughtcrime.securesms.megaphone.MegaphoneRepository}.
 */
public class MegaphoneDatabase extends SQLiteOpenHelper implements SignalDatabase {

  private static final String TAG = Log.tag(MegaphoneDatabase.class);

  private static final int    DATABASE_VERSION = 1;
  private static final String DATABASE_NAME    = "signal-megaphone.db";

  private static final String TABLE_NAME    = "megaphone";
  private static final String ID            = "_id";
  private static final String EVENT         = "event";
  private static final String SEEN_COUNT    = "seen_count";
  private static final String LAST_SEEN     = "last_seen";
  private static final String FIRST_VISIBLE = "first_visible";
  private static final String FINISHED      = "finished";

  public static final String CREATE_TABLE = "CREATE TABLE " + TABLE_NAME + "(" + ID            + " INTEGER PRIMARY KEY AUTOINCREMENT, " +
                                                                                 EVENT         + " TEXT UNIQUE, " +
                                                                                 SEEN_COUNT    + " INTEGER, " +
                                                                                 LAST_SEEN     + " INTEGER, " +
                                                                                 FIRST_VISIBLE + " INTEGER, " +
                                                                                 FINISHED      + " INTEGER)";

  private static volatile MegaphoneDatabase instance;

  private final Application    application;
  private final DatabaseSecret databaseSecret;

  public static @NonNull MegaphoneDatabase getInstance(@NonNull Application context) {
    if (instance == null) {
      synchronized (MegaphoneDatabase.class) {
        if (instance == null) {
          instance = new MegaphoneDatabase(context, DatabaseSecretProvider.getOrCreateDatabaseSecret(context));
        }
      }
    }
    return instance;
  }

  public MegaphoneDatabase(@NonNull Application application, @NonNull DatabaseSecret databaseSecret) {
    super(application, DATABASE_NAME, null, DATABASE_VERSION, new SqlCipherDatabaseHook());

    this.application    = application;
    this.databaseSecret = databaseSecret;
  }

  @Override
  public void onCreate(SQLiteDatabase db) {
    Log.i(TAG, "onCreate()");

    db.execSQL(CREATE_TABLE);

    if (DatabaseFactory.getInstance(application).hasTable("megaphone")) {
      Log.i(TAG, "Found old megaphone table. Migrating data.");
      migrateDataFromPreviousDatabase(DatabaseFactory.getInstance(application).getRawDatabase(), db);
    }
  }

  @Override
  public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
    Log.i(TAG, "onUpgrade(" + oldVersion + ", " + newVersion + ")");
  }

  @Override
  public void onOpen(SQLiteDatabase db) {
    Log.i(TAG, "onOpen()");

    if (DatabaseFactory.getInstance(application).hasTable("megaphone")) {
      Log.i(TAG, "Dropping original megaphone table from the main database.");
      DatabaseFactory.getInstance(application).getRawDatabase().rawExecSQL("DROP TABLE megaphone");
    }
  }

  public void insert(@NonNull Collection<Event> events) {
    SQLiteDatabase db = getWritableDatabase();

    db.beginTransaction();
    try {
      for (Event event : events) {
        ContentValues values = new ContentValues();
        values.put(EVENT, event.getKey());

        db.insertWithOnConflict(TABLE_NAME, null, values, SQLiteDatabase.CONFLICT_IGNORE);
      }

      db.setTransactionSuccessful();
    } finally {
      db.endTransaction();
    }
  }

  public @NonNull List<MegaphoneRecord> getAllAndDeleteMissing() {
    SQLiteDatabase        db      = getWritableDatabase();
    List<MegaphoneRecord> records = new ArrayList<>();

    db.beginTransaction();
    try {
      Set<String> missingKeys = new HashSet<>();

      try (Cursor cursor = db.query(TABLE_NAME, null, null, null, null, null, null)) {
        while (cursor != null && cursor.moveToNext()) {
          String  event        = cursor.getString(cursor.getColumnIndexOrThrow(EVENT));
          int     seenCount    = cursor.getInt(cursor.getColumnIndexOrThrow(SEEN_COUNT));
          long    lastSeen     = cursor.getLong(cursor.getColumnIndexOrThrow(LAST_SEEN));
          long    firstVisible = cursor.getLong(cursor.getColumnIndexOrThrow(FIRST_VISIBLE));
          boolean finished     = cursor.getInt(cursor.getColumnIndexOrThrow(FINISHED)) == 1;

          if (Event.hasKey(event)) {
            records.add(new MegaphoneRecord(Event.fromKey(event), seenCount, lastSeen, firstVisible, finished));
          } else {
            Log.w(TAG, "No in-app handing for event '" + event + "'! Deleting it from the database.");
            missingKeys.add(event);
          }
        }
      }

      for (String missing : missingKeys) {
        String   query = EVENT + " = ?";
        String[] args  = new String[]{missing};

        db.delete(TABLE_NAME, query, args);
      }

      db.setTransactionSuccessful();
    } finally {
      db.endTransaction();
    }

    return records;
  }

  public void markFirstVisible(@NonNull Event event, long time) {
    String   query = EVENT + " = ?";
    String[] args  = new String[]{event.getKey()};

    ContentValues values = new ContentValues();
    values.put(FIRST_VISIBLE, time);

    getWritableDatabase().update(TABLE_NAME, values, query, args);
  }

  public void markSeen(@NonNull Event event, int seenCount, long lastSeen) {
    String   query = EVENT + " = ?";
    String[] args  = new String[]{event.getKey()};

    ContentValues values = new ContentValues();
    values.put(SEEN_COUNT, seenCount);
    values.put(LAST_SEEN, lastSeen);

    getWritableDatabase().update(TABLE_NAME, values, query, args);
  }

  public void markFinished(@NonNull Event event) {
    String   query = EVENT + " = ?";
    String[] args  = new String[]{event.getKey()};

    ContentValues values = new ContentValues();
    values.put(FINISHED, 1);

    getWritableDatabase().update(TABLE_NAME, values, query, args);
  }

  public void delete(@NonNull Event event) {
    String   query = EVENT + " = ?";
    String[] args  = new String[]{event.getKey()};

    getWritableDatabase().delete(TABLE_NAME, query, args);
  }

  private @NonNull SQLiteDatabase getWritableDatabase() {
    return getWritableDatabase(databaseSecret.asString());
  }

  @Override
  public @NonNull SQLiteDatabase getSqlCipherDatabase() {
    return getWritableDatabase();
  }

  private static void migrateDataFromPreviousDatabase(@NonNull SQLiteDatabase oldDb, @NonNull SQLiteDatabase newDb) {
    try (Cursor cursor = oldDb.rawQuery("SELECT * FROM megaphone", null)) {
      while (cursor.moveToNext()) {
        ContentValues values = new ContentValues();

        values.put(EVENT, CursorUtil.requireString(cursor, "event"));
        values.put(SEEN_COUNT, CursorUtil.requireInt(cursor, "seen_count"));
        values.put(LAST_SEEN, CursorUtil.requireLong(cursor, "last_seen"));
        values.put(FIRST_VISIBLE, CursorUtil.requireLong(cursor, "first_visible"));
        values.put(FINISHED, CursorUtil.requireInt(cursor, "finished"));

        newDb.insert(TABLE_NAME, null, values);
      }
    }
  }
}

File: app/src/main/java/org/thoughtcrime/securesms/database/MegaphoneDatabase.kt
package org.thoughtcrime.securesms.database

import android.app.Application
import androidx.annotation.VisibleForTesting
import androidx.sqlite.db.SupportSQLiteDatabase
import net.zetetic.database.sqlcipher.SQLiteDatabase
import net.zetetic.database.sqlcipher.SQLiteOpenHelper
import org.signal.core.util.delete
import org.signal.core.util.forEach
import org.signal.core.util.insertInto
import org.signal.core.util.logging.Log
import org.signal.core.util.requireBoolean
import org.signal.core.util.requireInt
import org.signal.core.util.requireLong
import org.signal.core.util.requireNonNullString
import org.signal.core.util.select
import org.signal.core.util.update
import org.signal.core.util.withinTransaction
import org.thoughtcrime.securesms.crypto.DatabaseSecret
import org.thoughtcrime.securesms.crypto.DatabaseSecretProvider
import org.thoughtcrime.securesms.database.SqlCipherLibraryLoader.load
import org.thoughtcrime.securesms.database.model.MegaphoneRecord
import org.thoughtcrime.securesms.megaphone.Megaphones
import kotlin.concurrent.Volatile

/**
 * IMPORTANT: Writes should only be made through [org.thoughtcrime.securesms.megaphone.MegaphoneRepository].
 */
open class MegaphoneDatabase(
  application: Application,
  databaseSecret: DatabaseSecret
) : SQLiteOpenHelper(
  application,
  DATABASE_NAME,
  databaseSecret.asString(),
  null,
  DATABASE_VERSION,
  0,
  SqlCipherErrorHandler(application, DATABASE_NAME),
  SqlCipherDatabaseHook(),
  true
),
  SignalDatabaseOpenHelper {

  companion object {
    private val TAG = Log.tag(MegaphoneDatabase::class.java)

    private const val DATABASE_VERSION = 2
    private const val DATABASE_NAME = "signal-megaphone.db"

    private const val TABLE_NAME = "megaphone"
    private const val ID = "_id"

    /**
     * The event name, which is a key we use to tie it to views and whatnot.
     */
    private const val EVENT = "event"

    /**
     * How many times a megaphone was interacted with. This is most commonly the "snooze" count.
     */
    private const val INTERACTION_COUNT = "interaction_count"

    /**
     * The last time a megaphone was interacted with. This is most commonly the "snooze" timestamp.
     */
    private const val LAST_INTERACTION_TIMESTAMP = "last_interaction_timestamp"

    /**
     * The timestamp of when the megaphone was first shown to the user.
     */
    private const val FIRST_VISIBLE = "first_visible"

    /**
     * The timestamp of then when the last "view cycle" started. For instance, if a megaphone was
     * snoozed and then shown again, this will be the timestamp of when it was first shown again.
     * It is *not* updated every time a megaphone is seen, just at the start of the view cycle.
     * This is largely used to determine when to auto-snooze a megaphone.
     */
    private const val LAST_VISIBLE = "last_visible"

    /**
     * Whether a megaphone has been fully completed. When it's finished, it'll never be shown again.
     */
    private const val FINISHED = "finished"

    const val CREATE_TABLE: String = """CREATE TABLE $TABLE_NAME(
        $ID INTEGER PRIMARY KEY AUTOINCREMENT,
        $EVENT TEXT UNIQUE,
        $INTERACTION_COUNT INTEGER,
        $LAST_INTERACTION_TIMESTAMP INTEGER,
        $FIRST_VISIBLE INTEGER,
        $LAST_VISIBLE INTEGER DEFAULT 0,
        $FINISHED INTEGER
    )"""

    @Volatile
    private var instance: MegaphoneDatabase? = null

    @JvmStatic
    fun getInstance(context: Application): MegaphoneDatabase {
      if (instance == null) {
        synchronized(MegaphoneDatabase::class.java) {
          if (instance == null) {
            load()
            instance = MegaphoneDatabase(context, DatabaseSecretProvider.getOrCreateDatabaseSecret(context))
          }
        }
      }
      return instance!!
    }
  }

  @get:VisibleForTesting
  internal open val database: SupportSQLiteDatabase
    get() = writableDatabase

  override fun onCreate(db: SQLiteDatabase) {
    Log.i(TAG, "onCreate()")
    db.execSQL(CREATE_TABLE)
  }

  override fun onUpgrade(db: SQLiteDatabase?, oldVersion: Int, newVersion: Int) {
    Log.i(TAG, "onUpgrade($oldVersion, $newVersion)")

    if (oldVersion < 2) {
      db!!.execSQL("ALTER TABLE $TABLE_NAME ADD COLUMN $LAST_VISIBLE INTEGER DEFAULT 0")
      db.execSQL("ALTER TABLE $TABLE_NAME RENAME COLUMN seen_count TO interaction_count")
      db.execSQL("ALTER TABLE $TABLE_NAME RENAME COLUMN last_seen TO last_interaction_timestamp")
    }
  }

  override fun onOpen(db: SQLiteDatabase) {
    Log.i(TAG, "onOpen()")
    db.setForeignKeyConstraintsEnabled(true)
  }

  fun insert(events: Collection<Megaphones.Event>) {
    database.withinTransaction { db ->
      for (event in events) {
        db.insertInto(TABLE_NAME)
          .values(EVENT to event.key)
          .run(SQLiteDatabase.CONFLICT_IGNORE)
      }
    }
  }

  fun getAllAndDeleteMissing(): MutableList<MegaphoneRecord> {
    val records: MutableList<MegaphoneRecord> = mutableListOf()

    database.withinTransaction { db ->
      val missingKeys: MutableSet<String> = mutableSetOf()

      db.select()
        .from(TABLE_NAME)
        .run()
        .forEach { cursor ->
          val event = cursor.requireNonNullString(EVENT)
          val interactionCount = cursor.requireInt(INTERACTION_COUNT)
          val lastInteractionTime = cursor.requireLong(LAST_INTERACTION_TIMESTAMP)
          val firstVisible = cursor.requireLong(FIRST_VISIBLE)
          val lastVisible = cursor.requireLong(LAST_VISIBLE)
          val finished = cursor.requireBoolean(FINISHED)

          if (Megaphones.Event.hasKey(event)) {
            records += MegaphoneRecord(
              event = Megaphones.Event.fromKey(event),
              interactionCount = interactionCount,
              lastInteractionTime = lastInteractionTime,
              firstVisible = firstVisible,
              lastVisible = lastVisible,
              finished = finished
            )
          } else {
            Log.w(TAG, "No in-app handing for event '$event'! Deleting it from the database.")
            missingKeys += event
          }
        }

      for (missing in missingKeys) {
        db.delete(TABLE_NAME)
          .where("$EVENT = ?", missing)
          .run()
      }
    }

    return records
  }

  fun markFirstVisible(event: Megaphones.Event, time: Long) {
    database
      .update(TABLE_NAME)
      .values(FIRST_VISIBLE to time)
      .where("$EVENT = ?", event.key)
      .run()
  }

  fun markLastVisible(event: Megaphones.Event, time: Long) {
    database
      .update(TABLE_NAME)
      .values(LAST_VISIBLE to time)
      .where("$EVENT = ?", event.key)
      .run()
  }

  fun markInteractedWith(event: Megaphones.Event, interactionCount: Int, lastInteractionTimestamp: Long) {
    database
      .update(TABLE_NAME)
      .values(
        INTERACTION_COUNT to interactionCount,
        LAST_INTERACTION_TIMESTAMP to lastInteractionTimestamp,
        LAST_VISIBLE to 0L
      )
      .where("$EVENT = ?", event.key)
      .run()
  }

  fun markFinished(event: Megaphones.Event) {
    database
      .update(TABLE_NAME)
      .values(FINISHED to 1)
      .where("$EVENT = ?", event.key)
      .run()
  }

  fun delete(event: Megaphones.Event) {
    database
      .delete(TABLE_NAME)
      .where("$EVENT = ?", event.key)
      .run()
  }

  override fun getSqlCipherDatabase(): SQLiteDatabase {
    return writableDatabase
  }
}
